### PR TITLE
Add FSRS-7 + Logistic Regression Ensemble model

### PR DESCRIPTION
**Pull request by HuyAiT**
## Summary

Introduces a new ensemble algorithm (**FSRS-7-LR-Ensemble**) that blends FSRS-7 and Logistic Regression predictions in logit space, achieving strong results with only **72 parameters**.

- **Logit-space blending**: `w · logit_fsrs + (1-w) · logit_lr` with separate weights for same-day and non-same-day reviews
- **Per-user weight optimization**: Blending weights are fitted on a temporal holdout (last 30% of training data) using `scipy.optimize.minimize_scalar` with an L2 prior centered at 0.4
- **New LR feature**: `(1 - has_passed) × log1p(same_day_fails)` to capture pre-acquisition phase behavior
- **Utility additions**: `--min-user-id` CLI argument for partial benchmark runs, `gc.collect()` after each user

### Preliminary results (~450 users)

| Model | Params | LogLoss ↓ | RMSE(bins) ↓ | AUC ↑ |
|-------|--------|-----------|--------------|-------|
| FSRS-7 | 35 | 0.3264 | 0.0435 | 0.7179 |
| LogisticRegression | 35 | 0.3315 | 0.0514 | 0.6930 |
| **FSRS-7-LR-Ensemble** | **72** | **0.3237** | **0.0432** | **0.7176** |
| GRU-P | 297 | 0.3221 | 0.0414 | 0.7219 |

The ensemble ranks #1 among non-neural models and is competitive with much larger architectures.

Related issue: #331

## How to run

```bash
python script.py --algo FSRS-7-LR-Ensemble --short --secs --equalize_test_with_non_secs
```

## Test plan

- [x] Tested on ~450 users with `--short --secs --equalize_test_with_non_secs`
- [x] Verified no data leakage (temporal holdout for weight fitting, TimeSeriesSplit for evaluation)
- [x] Confirmed LR features work correctly with and without `--secs` flag
- [ ] Full 10k user benchmark (in progress)

### CODE REVIEW

File: models/fsrs7_lr_ensemble.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+        w_nsd = self.weights.get("w_fsrs_nsd", self.PRIOR[0])
+        w_sd = self.weights.get("w_fsrs_sd", self.PRIOR[1])
+        w = np.where(is_same_day, w_sd, w_nsd)
+        return w * logit_fsrs + (1.0 - w) * logit_lr
+
+    # ── training ─────────────────────────────────────────────────────────
+
+    def optimize(self, df: pd.DataFrame) -> Dict[str, Any]:
+        fsrs_model = FSRS7(config=self.fsrs_config)
+        from script import Trainer
```
[the remark this answers is not in the document]
Comment (HuyAiT) [reply models/fsrs7_lr_ensemble.py:87]: Fixed in d5e4e4e — replaced with `sys.modules["__main__"]` to reuse the already-loaded `Trainer` without re-executing `script.py` top-level code.

File: models/fsrs7_lr_ensemble.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+        sd_mask = sd_val.astype(bool)
+        nsd_mask = ~sd_mask
+
+        w_nsd = self.PRIOR[0]
+        if np.sum(nsd_mask) >= 10:
+            w_nsd = self._fit_weight(lf_val[nsd_mask], ll_val[nsd_mask], y_val[nsd_mask])
+
+        w_sd = self.PRIOR[1]
+        if np.sum(sd_mask) >= 10:
+            w_sd = self._fit_weight(lf_val[sd_mask], ll_val[sd_mask], y_val[sd_mask])
```
[the remark this answers is not in the document]
Comment (HuyAiT) [reply models/fsrs7_lr_ensemble.py:131]: This is a deliberate design choice, not an oversight. The blending weights (only 2 scalars) are heavily regularized with an L2 prior (`reg = 1/n`), which strongly shrinks them toward the prior of 0.4 on small datasets. Splitting the already-small per-user training data into disjoint sets for base model training vs. weight fitting would hurt base model performance significantly — each user typically has only a few hundred to a few thousand reviews.

The actual test set integrity is guaranteed by the benchmark's `TimeSeriesSplit`, which is completely separate from this internal holdout. The temporal holdout (last 30% of training data) provides ordering discipline for the weight fitting, and the L2 prior prevents the 2 scalar weights from overfitting to in-sample predictions.

File: models/fsrs7_lr_ensemble.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    def predict(self, df: pd.DataFrame) -> np.ndarray:
+        fsrs_model = FSRS7(config=self.fsrs_config, w=self.weights.get("fsrs_w"))
+        from utils import Collection
+
+        fsrs_ret, _, _ = Collection(fsrs_model, self.fsrs_config).batch_predict(df)
+
+        lr_model = LogisticRegression(
+            config=self.lr_config, state_dict=self.weights.get("lr_state")
+        )
+        lr_ret = lr_model.predict(df).cpu().numpy()
```
[the remark this answers is not in the document]
Comment (HuyAiT) [reply models/fsrs7_lr_ensemble.py]: Fixed in d5e4e4e — sub-models are now cached via `_get_sub_models()` and reused across partition calls. Cache is invalidated on `load_state_dict()`.

File: script.py [hunk 4]
```diff
@@ -466,8 +466,10 @@ def process(
     stats, raw = evaluate(
         y, p, save_tmp_df, config.get_evaluation_file_name(), user_id, config, w_list
     )
-    if config.model_name == "LogisticRegression" and model is not None:
+    if config.model_name in ["LogisticRegression", "FSRS-7-LR-Ensemble"] and model is not None:
         cast(Any, model).log(stats)
+    import gc
+    gc.collect()
```
[the remark this answers is not in the document]
Comment (HuyAiT) [reply script.py:472]: The ensemble trains two full models (FSRS-7 + LR) per user, which is significantly more memory-intensive than single-model runs. Without `gc.collect()`, memory accumulates across users when running with multiprocessing (8 workers), causing OOM on machines with limited RAM. The overhead of `gc.collect()` is negligible compared to the ~10-30s model training time per user.

That said, happy to move this to be ensemble-specific if preferred — e.g., only call it when `model_name == "FSRS-7-LR-Ensemble"`.

File: models/fsrs7_lr_ensemble.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (HuyAiT) [reply models/fsrs7_lr_ensemble.py:87]: Update: reverted to `from script import Trainer`. The `sys.modules["__main__"]` approach fails under multiprocessing `spawn` (child processes have `__mp_main__` instead of `__main__`). The `from script import Trainer` works correctly in both direct execution and spawned subprocesses — the top-level code (argument parsing) is idempotent and produces identical results.

**Comment by HuyAiT** [comment]
Closing this PR in favor of a cleaner replacement with the branch rebased on the latest main and the blending-weight fitting changed to use out-of-sample temporal holdout predictions.

Thanks for the earlier automated review feedback — it helped tighten the implementation.
